plugins/modules/cloudflare_dns.py

```diff
@@ -691,6 +691,10 @@ def ensure_dns_record(self, **kwargs):
                 "proto": params['proto'],
                 "service": params['service']
             }
+
+            if len(srv_data['name']) == 0:
+                srv_data['name'] = params['zone']  # Name is a required API field that needs to be set in case of a root record entry
```

Comment:
Do I understand you correctly that in case `params['zone'] == params['record']`, `params['zone']` must be used instead of `'@'`? Or why don't you use `'@'` when `srv_data['name']` is empty?

Reply:
See also my [other comment](); in the start of the module `@` is replaced with the `zone` name. It _does_ work with `@` (that is what i tested at first), but then the module is not idempotent anymore.

Imagine:
* Someone enters `record=@`, `zone=example.com` and `type=SRV` for this module
* In the start of the module `record=@` will be set to `record=example.com`. The record and zone are now equal.
* In the srv_data part the following line

Reply:
Hmm, but then how do you create a SRV entry for `record=example.com.example.com` when `zone=example.com`? The code will replace `example.com.example.com` with `example.com` and pass that on, and according to what you wrote it will then create a SRV entry for `example.com` and not for `example.com.example.com`. (That might be a problem of the API though :) )

Reply:
Hmm, `example.com.example.com` will indeed become `example.com`, resulting in a wrongly created SRV record.

It _is_ possibly to create a record, as you suggested, but that is more of a workaround than correct behavior probably. E.g. I misused the `proto` field for this.

```yaml
- zone: "example.com"
  record: "com.example.com"
  type: "SRV"
  ttl: 300
  service: "autodiscover"
  proto: "tcp.example"
  proxied: false
  priority: 0
  weight: 0
  port: 443
  value: "mail.example.com"
```

This results in the following API response (not everything is pasted here, and I didn't bother to replace my-domain with example.com):

```text
      "zone_name": "lenfer.ink",
      "name": "_autodiscover._tcp.lenfer.ink.lenfer.ink",
      "type": "SRV",
      "content": "0\t443\tmail.roylenferink.com",
      "priority": 0,
      "proxiable": false,
      "proxied": false,
      "ttl": 300,
      "locked": false,
      "data": {
        "name": "lenfer.ink.lenfer.ink",
        "port": 443,
        "priority": 0,
        "proto": "_tcp",
        "service": "_autodiscover",
        "target": "mail.roylenferink.com",
        "weight": 0
      },
```

As noticed; the proto is correctly set and the name results in `zone`.`zone`. So it might more be a limitation of the Ansible module than the Cloudflare API.

A 'normal' root SRV record (with service and proto set) looks like:

```text
      "zone_name": "lenfer.ink",
      "name": "_autodiscover._tcp.lenfer.ink",
      "type": "SRV",
      "content": "0\t443\tmail.roylenferink.com",
      "priority": 0,
      "proxiable": false,
      "proxied": false,
      "ttl": 300,
      "locked": false,
      "data": {
        "name": "lenfer.ink",
        "port": 443,
        "priority": 0,
        "proto": "_tcp",
        "service": "_autodiscover",
        "target": "mail.roylenferink.com",
        "weight": 0
      },
```

Reply:
The workaround looks ugly, but ... Anyway, let's merge this, since nobody brought up a better solution so far (and apparently nobody else wants to discuss this).

Reply:
I did some digging on why the `params['record'][:-len('.' + params['zone']` line was added (instead of just `params['record']`) and apparently it was merged in this PR: https://github.com/ansible/ansible-modules-extras/pull/1913

After removing the `[:-len('.' + params['zone']` part (and also the `if` statement I added in this PR) I _can_ set SRV records in an idempotent way (for both `@` and e.g. `example.com.example.com`). First time -> changed, subsequent runs -> ok.

I am a bit confused now on why [that specific commit](https://github.com/ansible/ansible-modules-extras/pull/1913/commits/7477fe51418dbb890faba8966282ab83e597e6af) was made.

Reply:
Hmm, without the `[:-len('.' + params['zone']` and the `if`, do you see any case which worked before that will stop working (or do something different)? The most important thing is that we don't break any existing (working) use-case.

Reply:
Sounds good to me!

The `if` statement for `@` might be needed for other record types though.

Reply:
> The `if` statement for `@` might be needed for other record types though.

I meant removing the check for `srv_data['name']` being empty (the initial intent of this PR); definitely _not_ removing the following part:

https://github.com/ansible-collections/community.general/blob/18a6bdd6aa09d0b67a9aa562c50402970f6d8c00/plugins/modules/cloudflare_dns.py#L411-L412

That is needed for all records indeed.

Reply:
Do you want to change this PR accordingly?